faker/providers/python/__init__.py

```diff
@@ -181,32 +191,67 @@ def pyfloat(
                 f"{sys.float_info.dig})",
             )
 
-        sign = ""
         if (min_value is not None) or (max_value is not None):
             # Copy values to ensure we're not modifying the original values and thus going out of bounds
-            left_min_value = min_value
-            left_max_value = max_value
+            temp_min_value = min_value
+            temp_max_value = max_value
             # Make sure left_digits still respected
-            if left_digits is not None:
-                if max_value is None:
-                    left_max_value = 10**left_digits  # minus smallest representable, adjusted later
-                if min_value is None:
-                    left_min_value = -(10**left_digits)  # plus smallest representable, adjusted later
+            if max_value is None:
+                temp_max_value = 10 ** (left_digits + right_digits)
+                if min_value is not None:
+                    temp_min_value = min_value * 10 ** right_digits
+            if min_value is None:
+                temp_min_value = -(10 ** (left_digits + right_digits))
+                if max_value is not None:
+                    if max_value < temp_min_value:
+                        temp_min_value = max_value - 1
+                    temp_max_value = max_value * 10 ** right_digits
 
             if max_value is not None and max_value < 0:
-                left_max_value += 1  # as the random_int will be generated up to max_value - 1
+                temp_max_value += 1  # as the random_int will be generated up to max_value - 1
             if min_value is not None and min_value < 0:
-                left_min_value += 1  # as we then append digits after the left_number
-            left_number = self._safe_random_int(
-                left_min_value,
-                left_max_value,
-                positive,
+                temp_min_value += 1  # as we then append digits after the left_number
+
+            opposite_signs = (temp_min_value * temp_max_value) < 0
+            if opposite_signs and left_digits_requested:
+                positive_number = self._safe_random_int(
+                    10 ** (left_digits + right_digits - 1),
+                    temp_max_value,
+                    positive=True,
+                )
+                negative_number = self._safe_random_int(
+                    temp_min_value,
+                    -(10 ** (left_digits + right_digits - 1)),
+                    positive=False,
+                )
+                number = self.random_element((positive_number, negative_number))
```

Comment:
This part is so that we can generate a number which respects these two constraints:
1. It must have a fixed amount of left digits
2. It can be either positive or negative.

In this case we need to choose two numbers, one in the negative side, and one in the positive. We then choose randomly between the two.